### PR TITLE
SSL support

### PR DESCRIPTION
**Pull request by cchantep**
Support SSL connection, when mongod is launched with `--sslMode requireSSL --sslPEMKeyFile my-cert.pem --sslAllowInvalidCertificates`

In connection options: `driver.connection(List("localhost:27017"), MongoConnectionOptions(sslEnabled = true, sslAllowsInvalidCert = true))`

### CODE REVIEW

**Comment by cchantep** [comment]
@sgodbillon Working but not to be merged right now:
- Depends on PR #277 (compat with MongoDB 3 stats)
- Want to a some related options in the configuration (e.g. `allowsInvalidCertificate` for self-signed cert)

File: driver/src/main/scala/core/nodeset.scala [hunk 1]
```diff
@@ -344,4 +354,12 @@ class ChannelFactory(options: MongoConnectionOptions, bossExecutor: Executor = E
     config.setConnectTimeoutMillis(options.connectTimeoutMS)
     channel
   }
+
+  private object TrustAnyManager extends javax.net.ssl.X509TrustManager {
+    import java.security.cert.X509Certificate
+
+    override def checkClientTrusted(cs: Array[X509Certificate], a: String) = {}
```
Comment (sgodbillon) [inline driver/src/main/scala/core/nodeset.scala]: I guess it's for your tests but we should avoid that.
The thing we can do is to add a config key to skip the certificate check (obviously set to false by default).

File: driver/src/test/resources/log4j2.xml [hunk 2]
```diff
@@ -1,14 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (sgodbillon) [inline driver/src/test/resources/log4j2.xml]: Was it necessary to delete this file?

Comment (cchantep) [inline driver/src/test/resources/log4j2.xml]: It should have been excluded. My bad.

File: driver/src/main/scala/core/nodeset.scala (hunk 1 above)
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: That's was I'm planning to do ;)

File: driver/src/main/scala/api/api.scala [hunk 3]
```diff
@@ -424,6 +424,8 @@ object MongoConnection {
         kv match {
           case ("authSource", v)           => unsupportedKeys -> result.copy(authSource = Some(v))
           case ("connectTimeoutMS", v)     => unsupportedKeys -> result.copy(connectTimeoutMS = v.toInt)
+          case ("sslEnabled", v)           => unsupportedKeys -> result.copy(sslEnabled = v.toBoolean)
```
Comment (cchantep) [inline driver/src/main/scala/api/api.scala]: New config options

File: driver/src/main/scala/core/actors.scala [hunk 4]
```diff
@@ -239,14 +239,14 @@ class MongoDBSystem(
       allChannelGroup(nodeSet).close.addListener(listener)
 
       // fail all requests waiting for a response
-      awaitingResponses.foreach( _._2.promise.failure(Exceptions.ClosedException) )
```
Comment (cchantep) [inline driver/src/main/scala/core/actors.scala]: Just code formatting

File: driver/src/main/scala/core/actors.scala [hunk 5]
```diff
@@ -333,17 +333,18 @@ class MongoDBSystem(
         } else true
       }
       if (!nodeSet.isReachable) {
-        if(nodeSetWasReachable) {
+        if (nodeSetWasReachable) {
```
Comment (cchantep) [inline driver/src/main/scala/core/actors.scala]: Just code formatting

File: driver/src/main/scala/core/actors.scala [hunk 6]
```diff
@@ -479,23 +480,20 @@ class MongoDBSystem(
           awaitingResponses -= response.header.responseTo
           if (response.error.isDefined) {
             logger.debug("{" + response.header.responseTo + "} sending a failure... (" + response.error.get + ")")
-            if (response.error.get.isNotAPrimaryError)
-              onPrimaryUnavailable()
+            if (response.error.get.isNotAPrimaryError) onPrimaryUnavailable()
```
Comment (cchantep) [inline driver/src/main/scala/core/actors.scala]: Just code formatting

File: driver/src/main/scala/core/actors.scala [hunk 7]
```diff
@@ -518,7 +516,8 @@ class MongoDBSystem(
               case IntField(_, v) => v != 0
               case DoubleField(_, v) => v != 0
             }.getOrElse(false)
-            if(processedOk) {
+
+            if (processedOk) {
```
Comment (cchantep) [inline driver/src/main/scala/core/actors.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 8]
```diff
@@ -61,16 +61,14 @@ case class NodeSet(
 
   def updateOrAddNode(ƒ: PartialFunction[Node, Node], default: Node) = {
     val (maybeUpdatedNodes, updated) = utils.update(nodes)(ƒ)
-    if (!updated)
-      copy(nodes = default +: nodes)
+    if (!updated) copy(nodes = default +: nodes)
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 9]
```diff
@@ -81,19 +79,18 @@ case class NodeSet(
   def updateByChannelId(id: Int)(ƒc: Connection => Connection)(ƒn: Node => Node) = {
     copy(nodes = nodes.map { node =>
       val (connections, updated) = utils.update(node.connections) {
-        case conn if {
-          conn.channel.getId == id
-        } => ƒc(conn)
+        case conn if (conn.channel.getId == id) => ƒc(conn)
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 10]
```diff
@@ -81,19 +79,18 @@ case class NodeSet(
...
-        ƒn(node.copy(connections = connections))
+      if (updated) ƒn(node.copy(connections = connections))
       else node
     })
   }
 
-  def pickByChannelId(id: Int): Option[(Node, Connection)] =
-    nodes.view.map(node => node -> node.connections.find(_.channel.getId() == id)).collectFirst {
-      case (node, connection) if connection.exists(_.status == ConnectionStatus.Connected) => node -> connection.get
+  def pickByChannelId(id: Int): Option[(Node, Connection)] =
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 11]
```diff
@@ -101,40 +98,34 @@ case class NodeSet(
       case (node, Some(connection)) => node -> connection
     }
 
-  private val pickConnectionAndFlatten: Option[Node] => Option[(Node, Connection)] = { node =>
-    node.map(node => node -> node.authenticatedConnections.pick).collect { case (node, Some(connection)) => (node, connection) }
+  private val pickConnectionAndFlatten: Option[Node] => Option[(Node, Connection)] = _.map(node => node -> node.authenticatedConnections.pick).collect {
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 12]
```diff
@@ -101,40 +98,34 @@ case class NodeSet(
...
   }
 
-  private def pickFromGroupWithFilter(roundRobiner: RoundRobiner[Node, Vector], filter: Option[BSONDocument => Boolean], fallback: => Option[Node]) = {
-    def nodeMatchesFilter(filter: BSONDocument => Boolean): Node => Boolean = _.tags.map(filter(_)).getOrElse(false)
-
-    filter.map { filter =>
-      roundRobiner.pickWithFilter(nodeMatchesFilter(filter))
-    }.getOrElse(fallback)
-  }
+  private def pickFromGroupWithFilter(roundRobiner: RoundRobiner[Node, Vector], filter: Option[BSONDocument => Boolean], fallback: => Option[Node]) =
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 13]
```diff
@@ -101,40 +98,34 @@ case class NodeSet(
...
       pickConnectionAndFlatten(mongos)
-    } else {
-      preference match {
-        case ReadPreference.Primary                    => pickConnectionAndFlatten(primary)
-        case ReadPreference.PrimaryPreferred(filter)   => pickConnectionAndFlatten(primary.orElse(pickFromGroupWithFilter(secondaries, filter, secondaries.pick)))
-        case ReadPreference.Secondary(filter)          => pickConnectionAndFlatten(pickFromGroupWithFilter(secondaries, filter, secondaries.pick))
-        case ReadPreference.SecondaryPreferred(filter) => pickConnectionAndFlatten(pickFromGroupWithFilter(secondaries, filter, secondaries.pick).orElse(primary))
-        case ReadPreference.Nearest(filter)            => pickConnectionAndFlatten(pickFromGroupWithFilter(nearestGroup, filter, nearest))
-      }
+    } else preference match {
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 14]
```diff
@@ -183,19 +174,20 @@ object ProtocolMetadata {
 }
 
 case class Connection(
-    channel: Channel,
-    status: ConnectionStatus,
-    authenticated: Set[Authenticated],
-    authenticating: Option[Authenticating]) {
+  channel: Channel,
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 15]
```diff
@@ -183,19 +174,20 @@ object ProtocolMetadata {
...
   def send(message: Request, writeConcern: Request) {
     channel.write(message)
     channel.write(writeConcern)
   }
-  def send(message: Request) {
-    channel.write(message)
-  }
 
-  def isAuthenticated(db: String, user: String) = authenticated.exists(auth => auth.user == user && auth.db == db)
+  def send(message: Request) = channel.write(message)
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 16]
```diff
@@ -278,13 +271,10 @@ class ContinuousIterator[A](iterable: Iterable[A], private var toDrop: Int = 0)
 
   val hasNext = iterator.hasNext
 
-  if (hasNext) {
-    drop(toDrop)
-  }
+  if (hasNext) drop(toDrop)
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 17]
```diff
@@ -310,17 +300,18 @@ class RoundRobiner[A, M[T] <: Iterable[T]](val subject: M[A], startAtIndex: Int
   private def pickWithFilter(filter: A => Boolean, tested: Int): Option[A] =
     if (length > 0 && tested < length) {
       val a = pick
-      if (!a.isDefined)
-        None
-      else if (filter(a.get))
-        a
+      if (!a.isDefined) None
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 18]
```diff
@@ -310,17 +300,18 @@ class RoundRobiner[A, M[T] <: Iterable[T]](val subject: M[A], startAtIndex: Int
...
+      else if (filter(a.get)) a
       else pickWithFilter(filter, tested + 1)
     } else None
 
-  def copy(subject: M[A], startAtIndex: Int = iterator.nextIndex) = new RoundRobiner(subject, startAtIndex)
+  def copy(subject: M[A], startAtIndex: Int = iterator.nextIndex) =
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: Just code formatting

File: driver/src/main/scala/core/nodeset.scala [hunk 19]
```diff
@@ -310,17 +300,18 @@ class RoundRobiner[A, M[T] <: Iterable[T]](val subject: M[A], startAtIndex: Int
...
+    new RoundRobiner(subject, startAtIndex)
 }
 
 class ChannelFactory(options: MongoConnectionOptions, bossExecutor: Executor = Executors.newCachedThreadPool, workerExecutor: Executor = Executors.newCachedThreadPool) {
+  import javax.net.ssl.{ KeyManager, SSLContext }
```
Comment (cchantep) [inline driver/src/main/scala/core/nodeset.scala]: SSL support

File: driver/src/main/scala/core/protocol/protocol.scala [hunk 20]
```diff
@@ -53,9 +53,7 @@ object `package` {
     }
 
     /** Write the contents of the given [[reactivemongo.core.protocol.ChannelBufferWritable]]. */
-    def write(writable: ChannelBufferWritable) {
-      writable writeTo buffer
-    }
+    def write(writable: ChannelBufferWritable) = writable writeTo buffer
```
Comment (cchantep) [inline driver/src/main/scala/core/protocol/protocol.scala]: Just code formatting

File: driver/src/main/scala/core/protocol/protocol.scala [hunk 21]
```diff
@@ -221,13 +219,12 @@ object Request {
    * @param op $op
    * @param documents $documentsA
    */
-  def apply(requestID: Int, responseTo: Int, op: RequestOp, documents: Array[Byte]): Request = {
```
Comment (cchantep) [inline driver/src/main/scala/core/protocol/protocol.scala]: Just code formatting

File: driver/src/main/scala/core/protocol/protocol.scala [hunk 22]
```diff
@@ -235,14 +232,18 @@ object Request {
    * @param op $op
    * @param documents $documentsA
    */
-  def apply(requestID: Int, op: RequestOp, documents: Array[Byte]): Request = Request.apply(requestID, 0, op, documents)
```
Comment (cchantep) [inline driver/src/main/scala/core/protocol/protocol.scala]: Just code formatting

File: driver/src/main/scala/core/protocol/protocol.scala [hunk 23]
```diff
@@ -235,14 +232,18 @@ object Request {
...
+  def apply(requestID: Int, op: RequestOp, documents: Array[Byte]): Request =
+    Request.apply(requestID, 0, op, documents)
+
   /**
    * Create a request.
    *
    * @param requestID $requestID
    * @param op $op
    */
-  def apply(requestID: Int, op: RequestOp): Request = Request.apply(requestID, op, new Array[Byte](0))
```
Comment (cchantep) [inline driver/src/main/scala/core/protocol/protocol.scala]: Just code formatting

File: driver/src/main/scala/core/protocol/protocol.scala [hunk 24]
```diff
@@ -286,19 +287,15 @@ object Response {
  * Response meta information.
  *
  * @param channelId the id of the channel that carried this response.
- * @param localAddress string representation of the local address of the channel
```
Comment (cchantep) [inline driver/src/main/scala/core/protocol/protocol.scala]: Remove doc for missing parameters

File: driver/src/main/scala/core/protocol/protocol.scala [hunk 25]
```diff
@@ -286,19 +287,15 @@ object Response {
...
-case class ResponseInfo(
-  channelId: Int)
-
+case class ResponseInfo(channelId: Int)
 
 sealed trait MongoWireVersion extends Ordered[MongoWireVersion] {
   def value: Int
 
   final def compare(x: MongoWireVersion): Int =
-    if(value == x.value) 0
```
Comment (cchantep) [inline driver/src/main/scala/core/protocol/protocol.scala]: Just code formatting

File: driver/src/test/scala/CollectionSpec.scala [hunk 26]
```diff
@@ -15,11 +14,13 @@ class CollectionSpec extends Specification with Tags {
 
   "ReactiveMongo" should {
     "create a collection" in {
-      Await.result(collection.create(), timeout) mustEqual (())
```
Comment (cchantep) [inline driver/src/test/scala/CollectionSpec.scala]: Just code formatting

**Comment by cchantep** [comment]
@sgodbillon This one is now rebased

**Pull request merged**